### PR TITLE
Make `from{,_mut}_ptr_range` const

### PR DESCRIPTION
**Pull request by WaffleLapkin**
This PR makes the following APIs `const`:
```rust
// core::slice

pub const unsafe fn from_ptr_range<'a, T>(range: Range<*const T>) -> &'a [T];
pub const unsafe fn from_mut_ptr_range<'a, T>(range: Range<*mut T>) -> &'a mut [T];
```

Tracking issue: #89792.
Feature for `from_ptr_range` as a `const fn`: `slice_from_ptr_range_const`.
Feature for `from_mut_ptr_range` as a `const fn`: `slice_from_mut_ptr_range_const`.

r? @oli-obk

### CODE REVIEW

**Comment by oli-obk** [comment]
@bors r-

Please add a test showing some uses of these functions from within const contexts

**Comment by WaffleLapkin** [comment]
@oli-obk should it be a documentation test or where should I place it?

**Comment by oli-obk** [comment]
Maybe wherever from_raw_parts is tested? I think there are some ui tests for when it errors and for when it succeeds

**Comment by WaffleLapkin** [comment]
I haven't found any. All tests that mention `slice::from_raw_parts` (`issue-76387-llvm-miscompile.rs`, `slice-of-zero-size-elements.rs`, `issue-76387.rs`, `const-size_of_val-align_of_val.rs`, `issue-91827-extern-types.rs`, `issue-36474.rs`, `method-mut-self-modifies-mut-slice-lvalue.rs`, `unsized3-rpass.rs`) seem to test something else.

**Comment by oli-obk** [comment]
That's not great, we should test the failure paths for all of these. Please add a new ui test that creates all kinds of bogus slices at compile time and puts them in the final value of a constant or static

File: src/test/ui/const-ptr/forbidden_slices.rs [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+    const_pointer_byte_offsets
+)]
+use std::{
+    mem::{size_of, MaybeUninit},
+    ptr,
+    slice::{from_ptr_range, from_raw_parts},
+};
+
+// Null is never valid for reads
+pub static S0: &[u32] = unsafe { from_raw_parts(ptr::null(), 0) };
```
[the remark this answers is not in the document]
Comment (oli-obk) [reply src/test/ui/const-ptr/forbidden_slices.rs:15]: oof. yea the compiler error will point into libcore instead. you can use an `// error-pattern:` comment at the top of the file

**Comment by WaffleLapkin** [comment]
I do not understand why CI is failing, locally the test passes just fine :(

**Comment by oli-obk** [comment]
The test is failing on 32 bit systems. You can either emit stderr per bitsize (there's a modifier comment for that) or disable the test on 32 bit. Either wfm, tho the former seems better for coverage. If you do the former you'll have to bless for 32 bit, too. You can do this by passing --target=foo to your x.py command with a 32 bit platform that you can build for locally (for me that is `i686-unknown-linux-gnu`). If you get random other test failures, those'll be run-pass tests that fail because you can't execute those (happens for me when testing run pass against an apple target)

**Comment by WaffleLapkin** [comment]
Again, I do not understand what's going on here :(

**Comment by oli-obk** [comment]
the test differences are in allocation ids. This is a problem across platforms, you can add

// normalize-stderr-test "alloc[0-9]+" -> "ALLOC_ID"

to the top of the test to replace all these ids in stderr with a canonical representation.

**Comment by oli-obk** [comment]
Hmm... not sure what makes your case special, but CI doesn't lie (I hope 😅)

@bors r+

**Comment by compiler-errors** [comment]
I think this failed in rollup again https://github.com/rust-lang/rust/pull/97561#issuecomment-1141523477 :sweat: 

@bors r- rollup=never

**Comment by WaffleLapkin** [comment]
>                 ╾─a79+0x1─╼ 04 00 00 00                         │ ╾──╼....

why is it just `a` in this case...

Anyway, I've normalized some more, maybe this time I've fixed all problems?

**Comment by compiler-errors** [comment]
Let's try it.

@bors r=oli-obk

**Comment by WaffleLapkin** [comment]
> warning: spurious network error (2 tries remaining): [28] Timeout was reached (download of `petgraph v0.5.1` failed to transfer more than 10 bytes in 30s)
> warning: spurious network error (2 tries remaining): [28] Timeout was reached (failed to download any data for `jemalloc-sys v0.5.0+5.3.0` within 30s)
> warning: spurious network error (2 tries remaining): [28] Timeout was reached (download of `scoped-tls v1.0.0` failed to transfer more than 10 bytes in 30s)
> warning: spurious network error (2 tries remaining): [28] Timeout was reached (failed to download any data for `rls-data v0.19.1` within 30s)
>  Downloaded rls-data v0.19.1
> warning: spurious network error (1 tries remaining): [28] Timeout was reached (download of `jemalloc-sys v0.5.0+5.3.0` failed to transfer more than 10 bytes in 30s)
> warning: spurious network error (1 tries remaining): [28] Timeout was reached (failed to download any data for `petgraph v0.5.1` within 30s)
> error: failed to download from `[https://crates.io/api/v1/crates/jemalloc-sys/0.5.0+5.3.0/download`](https://crates.io/api/v1/crates/jemalloc-sys/0.5.0+5.3.0/download%60)

**Comment by WaffleLapkin** [comment]
Yay 🎉

File: library/core/src/slice/raw.rs [hunk 2]
```diff
@@ -263,7 +264,8 @@ pub unsafe fn from_ptr_range<'a, T>(range: Range<*const T>) -> &'a [T] {
 ///
 /// [valid]: ptr#safety
 #[unstable(feature = "slice_from_ptr_range", issue = "89792")]
-pub unsafe fn from_mut_ptr_range<'a, T>(range: Range<*mut T>) -> &'a mut [T] {
+#[rustc_const_unstable(feature = "slice_from_mut_ptr_range_const", issue = "89792")]
```
Comment (WaffleLapkin) [inline library/core/src/slice/raw.rs:267]: Oh no, I messed up the feature gate

**Pull request merged**